lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+  }
+}
+
+/// Create the signature for the newly generated protocol constrained generic
+/// function.
+CanSILFunctionType
+ExistentialTransform::createExistentialSpecializedFunctionType() {
+  auto FTy = F->getLoweredFunctionType();
+  SILModule &M = F->getModule();
+  auto &Ctx = M.getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  auto OrigGenericSig = FTy->getGenericSignature();
+  /// First, add the old generic signature.
+  Builder.addGenericSignature(OrigGenericSig);
+
+  /// Convert existential argument types to generic argument types.
+  convertExistentialArgTypesToGenericArgTypes(Builder);
+
+  /// Compute the updated generic signature.
+  NewGenericSig = std::move(Builder).computeGenericSignature(
+      SourceLoc(), /*allowConcreteGenericParams=*/true);
+  NewGenericEnv = NewGenericSig->createGenericEnvironment();
+
+  /// Create a lambda for GenericParams.
+  auto getCanonicalType = [&](Type t) -> CanType {
+    return t->getCanonicalType(NewGenericSig);
+  };
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(FTy->getParameters().begin(), FTy->getParameters().end());
+
+  /// Create the complete list of parameters.
+  int Idx = 0;
+  llvm::SmallVector<SILParameterInfo, 8> InterfaceParams;
+  InterfaceParams.reserve(params.size());
+  for (auto &param : params) {
+    auto iter = Arg2GenericTypeMap.find(Idx);
+    if (iter != Arg2GenericTypeMap.end()) {
+      auto GenericParam = iter->second;
+      InterfaceParams.push_back(SILParameterInfo(getCanonicalType(GenericParam),
+                                                 param.getConvention()));
+    } else {
+      auto Ty = param.getType();
+      auto MappedTy = Ty;
+      if (!(Ty->hasTypeParameter() || Ty->hasArchetype())) {
+        MappedTy = Ty->mapTypeOutOfContext()->getCanonicalType();
+      }
+      InterfaceParams.push_back(
+          SILParameterInfo(MappedTy, param.getConvention()));
+    }
+    Idx++;
+  }
+
+  /// Now add the result type.
+  llvm::SmallVector<SILResultInfo, 8> InterfaceResults;
+  for (auto &result : FTy->getResults()) {
+    auto Ty = result.getType();
+    auto MappedTy = Ty;
+    if (!(Ty->hasTypeParameter() || Ty->hasArchetype())) {
+      MappedTy = Ty->mapTypeOutOfContext()->getCanonicalType();
+    }
+    auto InterfaceResult = result.getWithType(MappedTy);
+    InterfaceResults.push_back(InterfaceResult);
+  }
+
+  /// Now add the errors.
+  Optional<SILResultInfo> InterfaceErrorResult;
+  if (FTy->hasErrorResult()) {
+    auto errorResult = FTy->getErrorResult();
+    auto Ty = errorResult.getType();
+    auto MappedTy = Ty;
+    if (!(Ty->hasTypeParameter() || Ty->hasArchetype())) {
+      MappedTy = Ty->mapTypeOutOfContext()->getCanonicalType();
```

Comment:
If it doesn't have a type parameter, mapTypeOutOfContext() is a no-op. Remove the `if`